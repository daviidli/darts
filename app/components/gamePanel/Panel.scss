.waitingPanel {
	height: 100%;
	width: 100vw;
	background-color: rgba(0, 0, 0, 0.6);
	position: absolute;
	visibility: hidden;
	margin: 0 !important;

	&.visible {
		visibility: visible;
	}

	.clearDarts {
		font-weight: 500;
	}
}

.gamePanel {
	height: 100vh;

	.header {
		padding: 0 2em;

		.button {
			margin-left: 1em;
		}
	}

	.rounds {
		padding: 1em;
		padding-top: 2em;
		height: 75vh;
	}

	.center {
		height: 75vh;
	}

	.totalsContainer {
		display: flex;
		align-items: center;
		justify-content: center;
		height: 70%;
	}

	.totals {
		font-size: 14em;
		font-weight: 900;
		text-shadow: 0.06em 0.06em #e49708;
		text-align: center;
	}

	.dartsContainer {
		height: 30%;

		.dart {
			font-size: 3em;
			font-weight: 700;
			text-align: center;
			padding: 0.5em;
			width: 20em;

			&.active {
				background-color: rgba(51, 169, 255, 0.4);
				border-radius: 10em;
			}
		}
	}

	.players {
		height: 15vh;
		text-align: center;

		.player {
			background-color: rgba(0, 0, 0, 0.4);
			height: 100%;
			margin: 0 1em;
			max-width: 25em;

			&.active {
				background-color: rgba(249, 198, 96, 0.4);
			}
		}

		.playerTotal {
			font-weight: 700;
		}
	}
}
