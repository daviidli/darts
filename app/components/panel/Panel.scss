.gamePanel {
	height: 100vh;

	.center {
		height: 75vh;
	}
}
