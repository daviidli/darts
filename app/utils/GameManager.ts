/* eslint-disable no-await-in-loop */
import DartBoard from './dartBoard';
import store from '../store/store';
import {
	setCurrentRound,
	setCurrentPlayer,
	initializedRound,
	setWinner,
	setWaiting,
	setMaxDarts,
	setMaxRounds,
	setCurrentDart,
	setTotals
} from '../actions/gameActions';

abstract class GameManager {
	private dartboard: DartBoard;
	private running: boolean;
	private missedDart: boolean;
	private dartsCleared: boolean;

	constructor() {
		const state = store.getState();

		if (!state.serialPort.length) {
			throw new Error('No serial port selected.');
		}

		this.dartboard = new DartBoard(state.serialPort);
		this.running = false;
		this.missedDart = false;
		this.dartsCleared = false;
	}

	protected abstract initialize(): void;
	protected abstract async turn(): Promise<void>;

	protected async run() {
		const {
			maxRounds,
			maxDarts,
			players,
			clearDartsWaitTime
		} = store.getState();

		const roundsArr = this.createStrArray(
			players.length,
			maxRounds,
			maxDarts
		);
		store.dispatch(initializedRound(roundsArr));

		this.running = true;
		for (let r = 0; r < maxRounds && this.running; r++) {
			store.dispatch(setCurrentRound(r));
			for (let p = 0; p < players.length && this.running; p++) {
				store.dispatch(setCurrentPlayer(p));
				await this.turn();
				await this.waitForClearingDarts(clearDartsWaitTime);
			}
		}
	}

	protected async getDart(): Promise<string> {
		let currentDart;
		while (true) {
			try {
				currentDart = await this.dartboard.getResults(1, 500);
			} catch (e) {
				// continue regardless of error
				// todo: show error notification
			}

			if (this.missedDart) {
				this.missedDart = false;
				return 'miss';
			}

			if (currentDart) {
				return currentDart[0];
			}
		}
	}

	public start() {
		this.initialize();
		this.run();
	}

	public stop() {
		this.running = false;
		this.dartboard.close();
		store.dispatch(setMaxDarts(0));
		store.dispatch(setMaxRounds(0));
		store.dispatch(setCurrentPlayer(0));
		store.dispatch(setCurrentDart(0));
		store.dispatch(setCurrentRound(0));
		store.dispatch(initializedRound([]));
		store.dispatch(setTotals([]));
	}

	public setWinner(index: number) {
		this.stop();
		store.dispatch(setWinner(index));
	}

	public miss() {
		this.missedDart = true;
	}

	public getRunning() {
		return this.running;
	}

	public clearedDarts() {
		this.dartsCleared = true;
	}

	private async waitForClearingDarts(timeout = 0) {
		this.dartsCleared = false;
		store.dispatch(setWaiting(true));
		let timeouts = 0;

		while (true) {
			await this.timeout(1000);
			timeouts += 1;
			if (timeouts === timeout) {
				store.dispatch(setWaiting(false));
				return;
			}
			if (this.dartsCleared) {
				this.dartsCleared = false;
				store.dispatch(setWaiting(false));
				return;
			}
		}
	}

	private timeout(ms: number): Promise<void> {
		return new Promise(resolve => setTimeout(resolve, ms));
	}

	private createStrArray(
		iMax: number,
		jMax: number,
		kMax: number
	): string[][][] {
		const initialRounds: string[][][] = [];
		for (let i = 0; i < iMax; i++) {
			const arr = [];
			for (let j = 0; j < jMax; j++) {
				arr.push(new Array(kMax).fill(''));
			}
			initialRounds.push(arr);
		}
		return initialRounds;
	}
}

export default GameManager;
