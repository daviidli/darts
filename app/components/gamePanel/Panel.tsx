/* eslint-disable react/no-array-index-key */
import React from 'react';
import { Grid, Typography, Button } from '@material-ui/core';
import { connect } from 'react-redux';
import clsx from 'clsx';
import { useHistory } from 'react-router';
import { GiDart } from 'react-icons/gi';
import { stateType, playerType } from '../../reducers/types';
import GameManager from '../../utils/GameManager';
import Winner from '../../containers/Winner';
import styles from './Panel.scss';
import routes from '../../constants/routes.json';
import Rounds from '../rounds/Rounds';
import { getPreviousRounds, mappingToString } from '../../utils/dartUtils';
import DartText from '../dartText/DartText';

type Props = {
	totals: any[];
	rounds: string[][][];
	players: playerType[];
	currentPlayer: number;
	currentRound: number;
	currentDart: number;
	winner: number;
	maxRounds: number;
	waiting: boolean;
	gameManager: GameManager;
};

const Panel = (props: Props) => {
	const {
		totals,
		rounds,
		players,
		currentPlayer,
		currentRound,
		currentDart,
		winner,
		maxRounds,
		waiting,
		gameManager
	} = props;
	const history = useHistory();

	const goBack = () => {
		gameManager.stop();
		history.push(routes.HOME);
	};

	if (winner !== -1) {
		return <Winner />;
	}

	if (!gameManager.getRunning()) {
		return <div>starting up</div>;
	}

	const roundsShown = 6;
	return (
		<>
			<Grid
				container
				direction="column"
				justify="center"
				alignItems="center"
				spacing={8}
				className={clsx(styles.waitingPanel, {
					[styles.visible]: waiting
				})}
			>
				<Grid item>
					<Typography variant="h1" className={styles.clearDarts}>
						Clear Darts
					</Typography>
				</Grid>
				<Grid item>
					<Button
						variant="contained"
						color="secondary"
						onClick={() => gameManager.clearedDarts()}
					>
						Darts Cleared
					</Button>
				</Grid>
			</Grid>
			<Grid
				container
				direction="column"
				justify="center"
				alignItems="center"
				className={styles.gamePanel}
			>
				<Grid
					item
					container
					direction="row"
					justify="space-around"
					alignItems="center"
					className={styles.header}
				>
					<Grid item xs>
						<Button variant="outlined" onClick={goBack}>
							End Game
						</Button>
					</Grid>
					<Grid item xs container justify="flex-end">
						<Button
							variant="contained"
							color="secondary"
							onClick={() => gameManager.miss()}
						>
							Miss
						</Button>
					</Grid>
				</Grid>
				<Grid
					item
					container
					direction="row"
					justify="center"
					alignItems="center"
				>
					<Grid item xs={3} className={styles.rounds}>
						<Typography variant="h4">
							{`Rounds (${currentRound + 1}/${maxRounds})`}
						</Typography>
						<ol
							start={
								currentRound <= roundsShown
									? 1
									: currentRound - (roundsShown - 1)
							}
						>
							{getPreviousRounds(
								rounds[currentPlayer],
								roundsShown,
								currentRound
							).map((round: string[], i) => (
								<Rounds round={round} key={`round-${i}`} />
							))}
						</ol>
					</Grid>
					<Grid item xs={6} className={styles.center}>
						<div className={styles.totalsContainer}>
							<Typography variant="h2" className={styles.totals}>
								{totals[currentPlayer]}
							</Typography>
						</div>
						<Grid
							container
							direction="row"
							justify="center"
							alignItems="center"
							className={styles.dartsContainer}
						>
							{rounds[currentPlayer][currentRound].map(
								(dart, i) => {
									let content;

									if (dart === '') {
										content = <GiDart />;
									} else {
										content = (
											<DartText
												value={mappingToString(dart)}
											/>
										);
									}

									return (
										<Grid
											item
											xs
											key={`round-${i}`}
											className={clsx(styles.dart, {
												[styles.active]:
													i === currentDart
											})}
										>
											{content}
										</Grid>
									);
								}
							)}
						</Grid>
					</Grid>
					<Grid item xs={3}>
						{/* todo: add player info */}
					</Grid>
				</Grid>
				<Grid
					item
					container
					direction="row"
					justify="center"
					alignItems="center"
					className={styles.players}
				>
					{players.map((player, i) => (
						<Grid
							item
							xs
							container
							direction="column"
							justify="center"
							alignItems="center"
							key={player.id}
							className={clsx(styles.player, {
								[styles.active]: i === currentPlayer
							})}
						>
							<Grid item>
								<Typography
									variant="h2"
									className={styles.playerTotal}
								>
									{totals[i]}
								</Typography>
							</Grid>
							<Grid item>{player.name}</Grid>
						</Grid>
					))}
				</Grid>
			</Grid>
		</>
	);
};

const mapStateToProps = (state: stateType) => ({
	totals: state.totals,
	rounds: state.rounds,
	players: state.players,
	currentPlayer: state.currentPlayer,
	currentRound: state.currentRound,
	currentDart: state.currentDart,
	winner: state.winner,
	maxRounds: state.maxRounds,
	waiting: state.waiting
});

export default connect<any, any, any, stateType>(mapStateToProps, {})(Panel);
