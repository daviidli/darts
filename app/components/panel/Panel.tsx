import React from 'react';
import { Grid } from '@material-ui/core';
import GameManager from '../../utils/GameManager';
import styles from './Panel.scss';
import HeaderBar from '../headerBar/HeaderBar';
import PlayersPanel from '../../containers/PlayersPanel';
import RoundsPanel from '../../containers/RoundsPanel';
import DartsPanel from '../../containers/DartsPanel';

export type Props = {
	gameManager: GameManager;
};

const Panel = (props: Props) => {
	const { gameManager } = props;

	return (
		<Grid
			container
			direction="column"
			justify="center"
			alignItems="center"
			className={styles.gamePanel}
		>
			<Grid item>
				<HeaderBar gameManager={gameManager} />
			</Grid>
			<Grid
				item
				container
				direction="row"
				justify="center"
				alignItems="center"
			>
				<Grid item xs={3}>
					<RoundsPanel />
				</Grid>
				<Grid item xs={6} className={styles.center}>
					<DartsPanel gameManager={gameManager} />
				</Grid>
				<Grid item xs={3}>
					{/* todo: right side panel */}
				</Grid>
			</Grid>
			<Grid item>
				<PlayersPanel />
			</Grid>
		</Grid>
	);
};

export default Panel;
