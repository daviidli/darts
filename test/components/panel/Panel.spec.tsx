import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import Panel, { Props } from '../../../app/components/panel/Panel';
import GameManager from '../../../app/utils/GameManager';

Enzyme.configure({ adapter: new Adapter() });

const initialProps: Props = {
	gameManager: {} as GameManager
};

const setup = (props: Props = initialProps) => {
	const wrapper = shallow(<Panel {...props} />);
	return { wrapper };
};

describe('Panel component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
