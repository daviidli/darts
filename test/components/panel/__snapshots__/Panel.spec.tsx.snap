// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel component should match snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="gamePanel"
  container={true}
  direction="column"
  justify="center"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <HeaderBar
      gameManager={Object {}}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    direction="row"
    item={true}
    justify="center"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={3}
    >
      <Connect(RoundsPanelComponent) />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="center"
      item={true}
      xs={6}
    >
      <Connect(DartsPanelComponent)
        gameManager={Object {}}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={3}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
  >
    <Connect(PlayersPanelComponent) />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
